resources-service/src/resource/search/resource-cloudsearch-client.ts

```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+import {
+  CloudSearchDomainClient,
+  SearchCommand,
+  SearchCommandOutput,
+} from '@aws-sdk/client-cloudsearch-domain';
+// eslint-disable-next-line prettier/prettier
+import type { CloudSearchConfig } from '../../config/cloud-search';
+import { TermsAndFilters } from './search-types';
+import { AccountSID } from '@tech-matters/twilio-worker-auth';
+
+type SearchResultItem = {
+  id: string;
+  name?: string;
+  highlights: Record<string, string>;
+};
+
+type SearchResultSet = {
+  total: number;
+  items: SearchResultItem[];
+};
+
+const EMPTY_RESULT = { total: 0, items: [] };
+
+const client = (domainConfig: CloudSearchConfig) => {
+  const domainClient: CloudSearchDomainClient = new CloudSearchDomainClient({
+    endpoint: domainConfig.searchUrl.toString(),
+  });
+
+  return {
+    search: async (
+      accountSid: AccountSID,
+      { searchTermsByIndex }: TermsAndFilters,
+      start: number,
+      limit: number,
+    ): Promise<SearchResultSet> => {
+      // I'm a little concerned about how well sanitised this is. It might be possible to inject something
+      // However, since the query only returns names & IDs, and all we do is use the ID to look up from the DB...
+      // I'm not sure how much damage anyone could do by injecting something, beyond giving themselves an error message.
+      const queryClauses = Object.entries(searchTermsByIndex).flatMap(
+        ([index, { phrases, terms, weighting }]) => {
+          const phraseClauses = phrases.map(p => `(phrase field='${index}' boost=${weighting} '${p.replace(/'/g, `\\'`)}')`);
+          if (terms.length) {
+            return [
+              ...phraseClauses,
+              `(term field='${index}' boost=${weighting} '${terms.map(t=>t.replace(/'/g, `\\'`)).join(' ')}')`,
+            ];
+          }
+          return phraseClauses;
+        },
+      );
```

Comment:
I'd hope `@aws-sdk/client-cloudsearch-domain` (or something like it) would provide us tools for building these queries, to avoid injection attacks. But as you told, all they can get is a list of IDs and names so that wouldn't be terrible.

Reply:
Even the IDs and names would be requested via the resources service, and it would look up the resources, applying account & permissions (when we have any resource permissions) checks first, so they couldn't use it to get search results from other accounts.

Just something we have to keep an eye on as the system evolves I think, so we don't introduce something in the future where this possibility of injection attacks here becomes a real security issue.